### ISSUE TITLE
[Feature] Island bypass for custom fields

### ISSUE THREAD

**Issue by nitroSkirata** [Courseplay/Courseplay_FS22#3396]
**What is the problem this feature/enhancement solves?**
When you have maps with very wonly fields (Mecklemburg Vorpommern for example) Courseplay is not able to generate courses for some of the fields.
To remedy this, one can cut the field into smaller section, using customs fields.

Those fields with still contain islands though, so it would be awesome, if courseplay could generate courses on customs fields and still bypass islands.

**Describe the solution you'd like**
Enable island bypass for custom fields

**Describe alternatives you've considered**
cutting the field even further to avoid islands, but that would result in a lot of very small fields.

**Additional context**
![Screenshot 2024-09-03 182105](https://github.com/user-attachments/assets/4efc7481-bd7e-41bb-af6a-b0df5d7807c0)
Field 22 of the Mecklemburg map
Without island avoidance, even CP-5 and CP-6 would need to be further fragmented, to generate a collision free course.
With Island avoidance, it might even be possible to combine CP-1 and CP-2. Maybe even CP-5 and CP-6

**Comment by pvaiko** [comment]
I can't remember why we disabled island bypass for custom fields. It may have been the problem with detecting islands on a custom field, but not sure anymore.

**Comment by pvaiko** [comment]
(BTW, this isn't new, this was the same for the old course generator too...)

**Comment by nitroSkirata** [comment]
Can you check if it would break anything to reenable it?

Or quickly tell me how to enable it on my own, so I could check?

**Comment by pvaiko** [comment]
Replace 'false' here with 'true':
https://github.com/Courseplay/Courseplay_FS22/blob/dbe77e48261de28c2b8ac682e8638e53e343b9c1/scripts/ai/jobs/CpAIJobFieldWork.lua#L192

**Comment by nitroSkirata** [comment]
Tried it on a small scale, on the first few attempts, it works quite fine

**Comment by pvaiko** [comment]
@Tensuko  can you remember why we disabled island bypass on custom fields?

**Comment by pvaiko** [comment]
Oh, just remembered. Custom fields can be drawn anywhere, also areas which aren't fields (not plowed). In that case, the entire area would be detected as island.

**Comment by nitroSkirata** [comment]
> Oh, just remembered. Custom fields can be drawn anywhere, also areas which aren't fields (not plowed). In that case, the entire area would be detected as island.

Yep, thats it.

Even with island bypass disabled in the ingame generator settings, CP is not able to create a course on a non-Field with custom field island avoidance enabled

**Comment by Tensuko** [comment]
> @Tensuko can you remember why we disabled island bypass on custom fields?

I guess your 2nd answer is the reason.

What if we say the island needs its own custom "field" ?
So if we have a custom field within our custom field, the one within is counted as island?

I know that CAD software works like that for 3D construction, when you want a hole in a plate, you just draw one within your plate dimensions. But that would be the only way I can imange getting arround that non-field problem....

Else there is only left a "won't fix" / "can't do"...

**Comment by nitroSkirata** [comment]
> > @Tensuko can you remember why we disabled island bypass on custom fields?
> 
> I guess your 2nd answer is the reason.
> 
> What if we say the island needs its own custom "field" ? So if we have a custom field within our custom field, the one within is counted as island?
> 
> I know that CAD software works like that for 3D construction, when you want a hole in a plate, you just draw one within your plate dimensions. But that would be the only way I can imange getting arround that non-field problem....
> 
> Else there is only left a "won't fix" / "can't do"...

How about making the Island avoidance selection in the generator do the work?
IF it is selected to OFF, it is OFF for customs fields.
currently it seems to still run in the background.

**Comment by Tensuko** [comment]
> All the generator needs is a polygon for the field boundary and a list of other polygons which are the islands. So yes, this can easily be done, if you find a way on the UI to mark the islands...

Draw or record another custom field arround the islands. Is not that hard to do.
Could even make custom island for those who want that.
So I would say that is the best solution for custom fields.

As said, the Idea behind custom fields was for "non fields" wich would be big islands and it won't work at all.

**Comment by BoxerHUN** [comment]
> I can't remember why we disabled island bypass for custom fields. It may have been the problem with detecting islands on a custom field, but not sure anymore.


Hello.
I don't know what could have been the reason, but v7.4.0.0 did it properly, and v7.4.2.3 also does it after enabling bypass islands. He was tasked with 8 lands and 10 islands. I think he handles the islands well.

settings.bypassIslands:setValue(false) 
![v7 4 2 3_bypassislands_false](https://github.com/user-attachments/assets/c0a2bc57-4174-40f6-83f3-f67d916206ff)

settings.bypassIslands:setValue(true) 
![v7 4 2 3_bypassislands_true](https://github.com/user-attachments/assets/9ae73733-624c-45f8-9b70-23931cc776f3)

**Comment by nitroSkirata** [comment]
> > I can't remember why we disabled island bypass for custom fields. It may have been the problem with detecting islands on a custom field, but not sure anymore.
> 
> Hello. I don't know what could have been the reason, but v7.4.0.0 did it properly, and v7.4.2.3 also does it after enabling bypass islands. He was tasked with 8 lands and 10 islands. I think he handles the islands well.
> 
> settings.bypassIslands:setValue(false) ![v7 4 2 3_bypassislands_false](https://private-user-images.githubusercontent.com/45154190/376069407-c0a2bc57-4174-40f6-83f3-f67d916206ff.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4ODY4ODksIm5iZiI6MTcyODg4NjU4OSwicGF0aCI6Ii80NTE1NDE5MC8zNzYwNjk0MDctYzBhMmJjNTctNDE3NC00MGY2LTgzZjMtZjY3ZDkxNjIwNmZmLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE0VDA2MTYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzN2JhMjcxNGMwYzg4ZGFlYWI5OTI2Mjk0YjA0MTFhYWEwOGY5ZGRkOGFhMzdmMzZlZmJhNGNmYzdlOTE5NzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.i3JgoNF_eyr4EimMy3eBQPr5UZsf4MHIAQCYCTd5oN0)
> 
> settings.bypassIslands:setValue(true) ![v7 4 2 3_bypassislands_true](https://private-user-images.githubusercontent.com/45154190/376069408-9ae73733-624c-45f8-9b70-23931cc776f3.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4ODY4ODksIm5iZiI6MTcyODg4NjU4OSwicGF0aCI6Ii80NTE1NDE5MC8zNzYwNjk0MDgtOWFlNzM3MzMtNjI0Yy00NWY4LTliNzAtMjM5MzFjYzc3NmYzLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE0VDA2MTYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0Nzk5OTI0NDMyNzUxNGM0YmU3YzAzNmMzYTlkMjJkM2E2ODdhMmEwYmI0ZWU3OTFhYTNmOTU3NjNiMjExMGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FhB-ez1fMM-jp6gxsBQ1Ga_dRdzj2yIsRvTvtkaAL4w)

Yeah, it works well for this purpose, but it fails, if you create a field on non-farmland

**Comment by BoxerHUN** [comment]
Yeah, it works well for this purpose, but it fails, if you create a field on non-farmland

Ahh... I get it.

**Comment by Anonymous-any** [comment]
Hi, to continue in this theme been read the wiki so I continue here I will not read ALLL the issues !!
Ok my problem seem a bit same but not ^^.
I tried to DL the (True) version for testing from @nitroSkirata but I can't... It seem locked.

So I post my picture's, I try to make understand to CP there is a custom field CP-3 to avoid inside the custom field CP-2 because I put some factories and CP continue to generate into them... So what can I do for that avoid if there a way ?

I add you I been tried to create and make pause-continue when I created the previous CP-2 field related to same the path of the CP-3 but without the CP-3 (I hope you understand my bad eng
![4](https://github.com/user-attachments/assets/313b6532-6113-49c1-b0e2-dfc76c3ca092)
lish), then the line reached continue the (after paused) You know that CP do it, dot then the field CP-3 have meaning because the CP-2 only with CP-3 path simply not work to avoid, CP think its the same field... With the pictures maybe better understood Thank guys Regard
![4](https://github.com/user-attachments/assets/313b6532-6113-49c1-b0e2-dfc76c3ca092)
![1](https://github.com/user-attachments/assets/e895f591-5d3d-471e-b5a8-473645042adf)
![2](https://github.com/user-attachments/assets/35491ece-1f96-4dad-a84a-43f1b01af8ef)
![3](https://github.com/user-attachments/assets/ab1b36b9-92bb-410a-804f-931bd6b23422)

**Comment by Anonymous-any** [comment]
@nitroSkirata  'You gotta edit that one for yourself' I don't understand where you mean to edit what ? I can avoid this problem you say me ?
You see that colliders from factories CP have don't care about em :/ That mean it doesn't avoid them ! amazing...

'Island avoidance will always only avoid areas that are not field.'  <- Ok

**Comment by nitroSkirata** [comment]
@Anonymous-any On September 6th in this thread, pvaiko posted what needs to be changed, to enable Courseplay to avoid islands on custom fields.

This change is usefull, for when a field has a wonky shape, requiring it to be split up for courseplay to compute a course.

However, when this change is applied, coursplay won't be able to compute a course on a custom field, which ground is not considered farmland.

AKA custom fields can no longer be used to create new fields, or mow grass wich is not growing on farmland.

Neither of these options will cause courseplay to avoid an additional custom field painted on ground considered farmland.

If you really want CP to avoid that area, you need to create a boundry, by for example boxing it in with the paint and terraform tool in the ingame shop.

**Comment by Anonymous-any** [comment]
@nitroSkirata Thank for the return,
Ok I will try to use paint on ground then I hope that will be done without needed using terraform added !

But for the devs team :
Wouldn't it be simpler to add the collisions of the 'factories' object masks or items to be detected to just avoid them in my case without necessarily referring to the ground? Maybe it can make you think, I don't know. Thank you in any case for these details @nitroSkirata.
